modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala

```diff
@@ -36,28 +36,78 @@ import org.typelevel.log4cats.slf4j.Slf4jLogger
 
 object GlobalSnapshotStorage {
 
-  def make[F[_]: Async: KryoSerializer](
-    globalSnapshotLocalFileSystemStorage: GlobalSnapshotLocalFileSystemStorage[F],
-    inMemoryCapacity: NonNegLong
-  ): F[GlobalSnapshotStorage[F] with LatestBalances[F]] = {
+  private def makeResources[F[_]: Async]() = {
     def mkHeadRef = SignallingRef.of[F, Option[Signed[GlobalSnapshot]]](none)
     def mkOrdinalCache = MapRef.ofSingleImmutableMap[F, SnapshotOrdinal, Hash](Map.empty)
     def mkHashCache = MapRef.ofSingleImmutableMap[F, Hash, Signed[GlobalSnapshot]](Map.empty)
+    def mkNotPersistedCache = Ref.of(Set.empty[SnapshotOrdinal])
     def mkOffloadQueue = Queue.unbounded[F, SnapshotOrdinal]
 
     def mkLogger = Slf4jLogger.create[F]
 
-    mkLogger.flatMap { implicit logger =>
-      (mkHeadRef, mkOrdinalCache, mkHashCache, mkOffloadQueue).mapN {
-        make(_, _, _, _, globalSnapshotLocalFileSystemStorage, inMemoryCapacity)
-      }.flatten
+    (mkHeadRef, mkOrdinalCache, mkHashCache, mkNotPersistedCache, mkOffloadQueue, mkLogger).mapN {
+      (_, _, _, _, _, _)
     }
   }
 
+  def make[F[_]: Async: KryoSerializer](
+    globalSnapshotLocalFileSystemStorage: GlobalSnapshotLocalFileSystemStorage[F],
+    inMemoryCapacity: NonNegLong,
+    maybeRollbackHash: Option[Hash]
+  ): F[GlobalSnapshotStorage[F] with LatestBalances[F]] =
+    maybeRollbackHash match {
+      case Some(rollbackHash) =>
+        GlobalSnapshotStorage
+          .make[F](globalSnapshotLocalFileSystemStorage, inMemoryCapacity, rollbackHash)
+      case None =>
+        GlobalSnapshotStorage
+          .make[F](globalSnapshotLocalFileSystemStorage, inMemoryCapacity)
+    }
+
+  private def make[F[_]: Async: KryoSerializer](
+    globalSnapshotLocalFileSystemStorage: GlobalSnapshotLocalFileSystemStorage[F],
+    inMemoryCapacity: NonNegLong,
+    rollbackHash: Hash
+  ): F[GlobalSnapshotStorage[F] with LatestBalances[F]] =
+    globalSnapshotLocalFileSystemStorage
+      .read(rollbackHash)
+      .flatMap(ApplicativeError.liftFromOption(_, new Throwable("Rollback global snapshot not found!")))
+      .flatMap { rollbackSnapshot =>
+        makeResources().flatMap {
+          case (headRef, ordinalCache, hashCache, notPersistedCache, offloadQueue, logger) =>
+            implicit val l = logger
+
+            headRef.set(rollbackSnapshot.some) >>
+              ordinalCache(rollbackSnapshot.ordinal).set(rollbackHash.some) >>
+              hashCache(rollbackHash).set(rollbackSnapshot.some) >>
```

Comment:
Instead of that, you can just `prepend` the snapshot in the same way we prepend the genesis 🤔